src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java

```diff
@@ -134,12 +136,14 @@ public MesosCloud(
     this.slaveInfos = slaveInfos;
     this.checkpoint = checkpoint;
     this.onDemandRegistration = onDemandRegistration;
-
-    JenkinsScheduler.SUPERVISOR_LOCK.lock();
-    try {
-      restartMesos();
-    } finally {
-      JenkinsScheduler.SUPERVISOR_LOCK.unlock();
+    this.setJenkinsURL(jenkinsURL);
+    if(!onDemandRegistration) {
```

Comment:
Bug was this: Even If  'ondemand' registration is selected, Mesos plugin would connect(restartMesos) to Mesos master when the user clicks save in Jenkins global configuration

So, I just included the check if the user has enabled ondemand registration before restarting Schedular.